File: Objective-C/LRI/Extensions/UIImageView+Extension.m
//
//  UIImageView+Extension.m
//  LRI
//
//  Created by Development on 05.07.2022.
//

#import <UIKit/UIKit.h>

@interface ImageCache : NSObject
+ (instancetype)shared;
@property (nonatomic, strong) NSCache *cache;
@end

@implementation ImageCache

static ImageCache *shared = nil;

+ (instancetype)shared {
    if (shared == nil) {
        shared = [[ImageCache alloc] init];
    }
    return shared;
}

- (UIImage *)getObjectForKey:(NSString *)key {
    return  [self.cache objectForKey:key];
}

- (void)setObject:(UIImage *)object forKey:(NSString *)key {
    [self.cache setObject:object forKey:key];
}

@end
 
@implementation UIImageView (Extension)

- (void)asyncImageWith:(NSString *) urlString
       withPlaceholder:(UIImage *) placeholder
        withCompletion:(void (^)(UIImage *image, BOOL wasCached)) completion {
    self.image = placeholder;
    
    UIImage *image = [ImageCache.shared getObjectForKey:urlString];
    if (image != nil) {
        completion(image, true);
        return;
    }
    
    NSURL *url = [NSURL URLWithString:urlString];
    NSURLSessionDataTask *task = [NSURLSession.sharedSession dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if ((data != nil) && (error == nil)) {
                UIImage *image = [UIImage imageWithData:data];
                completion(image, false);
            }
        });
    }];
    [task resume];
}

@end

File: SwiftUI/LRI/Extensions/UIImageView+Extension.swift
//
//  UIImageView+Extension.swift
//  LRI
//
//  Created by Development on 02.07.2022.
//

import Foundation
import UIKit

class ImageCache {
    static let shared = ImageCache()
    private let cache = NSCache<NSString, UIImage>()
    func getObject(forKey key: String) -> UIImage? {
        cache.object(forKey: NSString(string: key))
    }
    func setObject(_ object: UIImage, forKey key: String) {
        cache.setObject(object, forKey: NSString(string: key))
    }
}

extension UIImageView {
    
    func asyncImage(with urlString: String,
                    placeholder: UIImage? = UIImage(named: "placeholder"),
                    wasFetchedByCache: @escaping (UIImage, Bool) -> Void) {
        self.image = placeholder
        
        if let image = ImageCache.shared.getObject(forKey: urlString)  {
            // image was fecthed by cache in memory
            wasFetchedByCache(image, true)
            return
        }
        
        guard let url = URL(string: urlString) else { return }
        let task = URLSession.shared.dataTask(with: url) { data, _, _ in
            DispatchQueue.main.async {
                if let data = data, let image = UIImage(data: data) {
                    ImageCache.shared.setObject(image, forKey: urlString)
                    // image was fecthed by api request
                    wasFetchedByCache(image, false)
                }
            }
        }
        task.resume()
    }

}
